### ISSUE TITLE
Dataset hyphenation

### ISSUE THREAD

**Issue by lostanlen** [mir-dataset-loaders/mirdata#210]
Spotted during CR of #207 

Datasets such as `RWC-Popular`, `RWC-Classical` have a hyphen in their `DATASET_DIR` string, but have a space in the dataset table (#203)

What should be do about this inconsistency?

**Comment by kyungyunlee** [comment]
Looking at this [doc]( https://github.com/mir-dataset-loaders/mirdata/blob/master/docs/source/datasets.rst ), naming pattern such as name <=> DATASET_DIR and module <=> mirdata.dataset_name  would be helpful.

**Comment by lostanlen** [comment]
@kyungyunlee Exactly. We want a two-column associative table:
rwc_classical <-> "RWC Classical"
groove_midi <-> "Groove MIDI"
medley_solos_db <-> "Medley-solos-DB" (this one does has hyphens natively!)
etc.
But I would rather having up to _three_ different variants of the same name for every dataset.

**Comment by rabitt** [comment]
@lostanlen does DATASET_DIR have to be exactly equivalent to the official name of the dataset? (e.g. the default folder for the dataset RWC Classical is `RWC-Classical`) 

Note that the table you mention does not list the value for DATASET_DIR at all - just the name of the dataset!

That said, and the list of datasets table aside, I'm in favor of having a standard mapping between module (mirdata.dataset_name) and DATASET_DIR.

**Comment by lostanlen** [comment]
> @lostanlen does DATASET_DIR have to be exactly equivalent to the official name of the dataset? 

I think so. I see no reason for having a folder name that is different from the official name.

> That said, and the list of datasets table aside, I'm in favor of having a standard mapping between module (mirdata.dataset_name) and DATASET_DIR.

I agree. How about this?

```import re

def name_to_module(name):
    return re.sub('[^0-9a-z]+', '_', name.lower())

names = """Beatles
DALI
GuitarSet
GTZAN genre
Ikala
MedleyDB Melody
MedleyDB Pitch
Medley-solos-DB
ORCHSET
RWC Classical
RWC Jazz
RWC Popular
Salami
TinySOL""".split("\n")

print("\n".join(map(name_to_module, names)))
```

Result:
```
beatles
dali
guitarset
gtzan_genre
ikala
medleydb_melody
medleydb_pitch
medley_solos_db
orchset
rwc_classical
rwc_jazz
rwc_popular
salami
tinysol
```

**Comment by rabitt** [comment]
> I think so. I see no reason for having a folder name that is different from the official name.

I have one important reason - spaces in filepaths = bad and annoying :(

For me, the original name of the dataset is not part of the code itself anywhere - it's only part of documentation. In the code, we can choose a convention for DATASET_DIR and the name of the module, and this can easily be adjusted in a non-uniform way from the original name.

**Comment by rabitt** [comment]
> ```
> names = """Beatles
> DALI
> GuitarSet
> GTZAN genre
> Ikala
> MedleyDB Melody
> MedleyDB Pitch
> Medley-solos-DB
> ORCHSET
> RWC Classical
> RWC Jazz
> RWC Popular
> Salami
> TinySOL""".split("\n")
> ```

note that also, this string you've defined does not appear anywhere in the code, because the API only relies on module_name and DATASET_DIR.

**Comment by lostanlen** [comment]
@rabitt That's true. I copy-pasted the string from `README.md`

>  I have one important reason - spaces in filepaths = bad and annoying :(

I see two options:
(1) folder name = module name
(2) folder name = official name, with " " replaced by "-". But that means we need to maintain a list of official names somewhere.
(3) folder name = name as it appears on mir-datasets.yaml (ties down to #152 )

**Comment by rabitt** [comment]
@lostanlen I'd support either option (1) or (3)! One concern I have is backwards compatibility - if we change the DATASET_DIR values it will mean that those with previously downloaded datasets in an old version of the folder name will suddenly have mirdata pointing to nonexistent files. Maybe we can start by agreeing on a convention moving forward?

**Comment by magdalenafuentes** [comment]
+1 on convention moving forward, don't like the idea of not being backwards compatible because of this.

**Comment by lostanlen** [comment]
@rabitt I could have it either way.

 Option (1) is by far easiest to enforce programmatically
```python
DATASET_DIR = os.path.splitext(__file__)[0])
```
 but it will look a little bit strange to newcomers. And it breaks compatibility for all datasets.

Option (3) assumes that we have solved #152, so it will require some coordination with Alexander Lerch et al. It can potentially made backwards compatible, but it demands a greater deal of bookkepping.

**Comment by rabitt** [comment]
For readability, I prefer if `DATASET_DIR` is explicitly defined in the code, but we can add a test for all future modules in `test_loaders.py` that `DATASET_DIR` = cased module name to enforce the convention in the future.

**Comment by lostanlen** [comment]
I made a script to compare name in table, name of directory (`DATASET_DIR`), name in docstring/readme, and name of submodule:

```
df = pd.DataFrame(columns=["table", "directory", "docstring", "submodule"])
table = """The Beatles Dataset
DALI
Gtzan-Genre
GuitarSet
iKala
Medley-solos-DB
MedleyDB Melody Subset
MedleyDB Pitch Tracking Subset
Orchset
RWC Classical
RWC Jazz
RWC Pop
Salami
TinySOL""".split("\n")

for submodule_id, submodule_str in enumerate(sorted(mirdata.__all__)):
    module_str = "mirdata." + submodule_str
    submodule = importlib.import_module(module_str)
    docstring = submodule.__doc__.split(" Loader")[0].split(" Dataset")[0]
    row = {
        "table": table[submodule_id],
        "submodule": submodule_str,
        "directory": submodule.DATASET_DIR,
        "docstring": docstring,
    }
    df = df.append(row, ignore_index=True)
```

![Screen Shot 2020-04-07 at 2 38 40 PM](https://user-images.githubusercontent.com/3943142/78670296-d1bab080-78dd-11ea-9dcf-367dfb571a85.png)

**Comment by lostanlen** [comment]
From the above, it appears that the manual input should perhaps be ... the docstring!

docstring to table (in the scope of #166): repeat verbatim
docstring to directory: replace spaces by hyphens
directory to submodule name: convert to lowercase, replace hyphens by underscores

This can be unit-tested. It's also backwards compatible in terms of module names and folder names on the filesystem. The way i obtained the docstring is a bit of a hack, but it works.

**Comment by rabitt** [comment]
> docstring to table (in the scope of #166): repeat verbatim
docstring to directory: replace spaces by hyphens
directory to submodule name: convert to lowercase, replace hyphens by underscores

Are you suggesting that we programatically set the table name, DATASET_DIR value, and submodule name based on the docstring? Or are you proposing this is the convention we adopt?

**Comment by lostanlen** [comment]
> Are you suggesting that we programatically set the table name, DATASET_DIR value, and submodule name based on the docstring? Or are you proposing this is the convention we adopt?

@rabitt at the end of the day, i don't mind if this is manual or automated. i was just showing a script to visualize potential inconsistencies. the less manual input, the less error-prone

**Comment by lostanlen** [comment]
> For readability, I prefer if DATASET_DIR is explicitly defined in the code, but we can add a test for all future modules in test_loaders.py that DATASET_DIR = cased module name to enforce the convention in the future.

Yes, let's do this. (i love PR #105)

**Comment by rabitt** [comment]
This is fixed in #296 !
